src/lib/mac/kmac/kmac.cpp

```diff
@@ -0,0 +1,224 @@
+      bytes_appended++;
+      }
+   if(!IS_LEFT_ENCODE)
+      {
+      output_container.push_back(i);
+      bytes_appended++;
+      }
+   return bytes_appended;
+   }
+
+template <typename T>
+size_t left_encode(size_t s, T& output_container)
+   {
+   return left_or_right_encode<true>(s, output_container);
+   }
+
+template <typename T>
+size_t right_encode(size_t s, T& output_container)
+   {
+   size_t result = left_or_right_encode<false>(s, output_container);
+   return result;
+   }
+
+
+
+
+size_t byte_len_from_bit_len(size_t bit_length)
+   {
+   if(bit_length % 8)
+      {
+      throw Invalid_Argument("cannot convert byte length to bit length that is not a multiple of 8");
+      }
+   return bit_length / 8;
+   }
+
+size_t bit_len_from_byte_len(size_t byte_length)
+   {
+
+   size_t bit_length = 8*byte_length;
+   if(bit_length < byte_length)
+      {
+      throw Botan::Invalid_Argument("byte length is too large. Only byte lengths of up to "
+                                    + std::to_string(std::numeric_limits<size_t>::max() / 8) + " are supported on this platform in this function.");
+      }
+   return bit_length;
+   }
+
+template <typename T>
+void encode_string(const uint8_t* input, size_t input_byte_length, T& output_container)
+   {
+   // left_encode(*bitlen* of input)
+   size_t written = left_encode(bit_len_from_byte_len(input_byte_length), output_container);
+   output_container.insert(output_container.end(), input, &input[input_byte_length]);
+   written += input_byte_length;
+   }
+
+template <typename T>
+void byte_pad(uint8_t input[], size_t input_byte_length, size_t w_in_bytes, T& output_container)
+   {
+   size_t written_bytes = left_encode(w_in_bytes, output_container);
+   output_container.insert(output_container.end(), input, &input[input_byte_length]);
+   written_bytes += input_byte_length;
+   if(w_in_bytes > written_bytes)
+      {
+      size_t nb_trail_zeroes = w_in_bytes - written_bytes;
+      std::vector<uint8_t> trailing_zeroes(nb_trail_zeroes, 0);
+      output_container.insert(output_container.end(), &trailing_zeroes[0], &trailing_zeroes[trailing_zeroes.size()]);
+      }
+
+   }
+
+}
+
+void KMAC256::clear()
+   {
+   zap(m_key);
+   m_key_set = false;
+   m_hash.clear();
+   }
+std::string KMAC256::name() const
+   {
+   return std::string("KMAC256(" + std::to_string(m_output_bit_length) + ")");
+   }
+std::unique_ptr<MessageAuthenticationCode> KMAC256::new_object() const
+   {
+   return std::make_unique<KMAC256>(m_output_bit_length);
+   }
+
+size_t KMAC256::output_length() const
+   {
+   return m_output_bit_length/8;
+   }
+
+Key_Length_Specification KMAC256::key_spec() const
+   {
+   // KMAC supports key lengths from zero up to 2²⁰⁴⁰ (2^(2040)) bits
+   // https://nvlpubs.nist.gov/nistpubs/specialpublications/nist.sp.800-185.pdf#page=28
+   return Key_Length_Specification(0, std::numeric_limits<size_t>::max());
```

Comment:
There is no practical reason to support keys larger than 512 bits.

This will just cause weird confusion (and out of memory errors) when someone tries to use `mac->maximum_keylength()`

Reply:
If 3.0 wouldn't have been out yet, I would have suggested making the maximum value a `std::optional<>` to clearly communicate "no relevant upper limit in practice".

Maybe provide a "smallish" upper bound but allow to use KMAC with larger keys? This would require a way to pass `Symmetric_Algorithm::valid_keylength()` for larger keys than `Key_Length_Spec` advertises in that key. That makes me feel uneasy as will. 🤔